Replace activerecord-import gem for Rails 6

kevindew opened the pull request:
This gem is no longer necessary since ActiveRecord 6 was released with
the [#insert_all method][1]. The ActiveRecord#insert_all method is lower
level than the gem so we have some extra considerations. These are:

1. We lose the automatic setting of timestamps so have to specify these
   manually
2. Validations will no longer run on models created in bulk, this is
   arguably a win as these don't have any user input and it seems
   unnecessary to hydrate thousands of models to perform the exact same
   migrations
3. We can no longer pass in an empty array of records to create a no-op,
   ActiveRecord will complain with an exception. Instead we need a guard
   clause.

More info in the commits - but they are a bit repetitive.

https://blog.bigbinary.com/2019/04/15/bulk-insert-support-in-rails-6.html

benthorner commented on app/models/subscription_content.rb:
```diff
@@ -1,25 +1,12 @@
-
-  validate :presence_of_content_change_or_message
-
-  def presence_of_content_change_or_message
```
Or....[this beauty](https://github.com/alphagov/service-manual-publisher/pull/816/commits/4eff0b9afbcc70ae75b376329725c5e45255e19d) (you'll need to scroll the not-quite-deleted diff at the end to see).

benthorner commented on app/models/email.rb:
```diff
@@ -15,8 +15,6 @@ class Email < ApplicationRecord
   enum status: { pending: 0, sent: 1, failed: 2 }
   enum failure_reason: { permanent_failure: 0, retries_exhausted_failure: 1, technical_failure: 2 }
 
-  validates :address, :subject, :body, presence: true
```
Minor: you could also do this for [DigestRunSubscriber](https://github.com/alphagov/email-alert-api/blob/master/app/models/digest_run_subscriber.rb), some columns of [SubscriberList](https://github.com/alphagov/email-alert-api/blob/master/app/models/subscriber_list.rb#L11) and some of the ones for [Message](https://github.com/alphagov/email-alert-api/blob/master/app/models/message.rb#L7), which are weirdly self-contradictory (is it blank or is it not!).

kevindew replied:
Oh cool - will take a look at what can be done there.

kevindew replied:
I've removed the DigestRunSubscriber ones.

SubscriberList we'd still want them as this a record that is created by a HTTP request so we'd want to surface ActiveRecord errors as a 422 response - compared to the 500 response you'd get from the PostgreSQL constraints.

Same situation about Message, though you're right the phrasing seems super unintuitive with that validation.  What it actually means is that it'll only apply the CriteriaSchemaValidator when the input is not blank, which is useful so we don't have to check it's present in the validator - but doesn't come out very logical.

benthorner replied:
Oh, I seee. Thanks for explaining!

benthorner commented on app/builders/content_change_email_builder.rb:
```diff
@@ -8,11 +8,10 @@ def self.call(*args)
-      records,
-      ProcessAndGenerateEmailsWorker::BATCH_SIZE,
-    )
+    return [] if records.empty?
```
Ah, I see you explain the PR comments. Since we've got a couple of repeated pieces of knowledge (timestamps and empty list handling), I'm wondering if we should have some kind of wrapper/helper method to deal with those for us. I realise we're also trying to move away from the custom stuff, so we could also just add comments or ensure the behaviour is covered in tests; as-is, it looks kind of optional.

kevindew replied:
We should be able to remove this conditional soon based on my other PR where we have a guard clause before this: https://github.com/alphagov/email-alert-api/commit/48ecf33e0854a7e448daf28196556507013940bb#diff-7fbb3f7a42ce78544138fb6b12302791R44

Sorry I forgot to add this info on the commit message for this. Will amend

benthorner replied:
Thanks. I notice there are two other occurrences (e.g. [here](https://github.com/alphagov/email-alert-api/pull/1288/files#diff-68c9c8c4027cb03116b9b692da62b20eR13)) that might live for longer, and could lead to an error in production if someone felt the empty array check was unnecessary in future. Worth a test?

kevindew replied:
Oh sure, just added tests for those.

kevindew commented on app/workers/process_message_and_generate_emails_worker.rb:
```diff
@@ -30,12 +30,18 @@ def perform(message_id)
-        batch_size: 500,
-      )
+      subscription_ids(message).each_slice(500) do |group|
+        now = Time.zone.now
```
I've got a model approach looming for this one in the process PR: https://github.com/alphagov/email-alert-api/commit/48ecf33e0854a7e448daf28196556507013940bb#diff-3c013211d8ee5dfbed7aff67d0df03d8R26-R45

I'll keep an eye on the other changes to see where some feel more like they should be done at the model level.

kevindew commented on spec/models/subscription_content_spec.rb:
```diff
@@ -26,40 +26,4 @@
     end
   end
-
-  describe ".import!" do
```
In most of the tests we cover this with tests such as: 

https://github.com/alphagov/email-alert-api/blob/534938ef180b6ef41b76c1d05107707bfa7f7dfc/spec/workers/process_content_change_and_generate_emails_worker_spec.rb#L27-L34

benthorner replied:
Oh yes, thanks for the pointer to those 👍.

kevindew commented on app/services/matched_content_change_generation_service.rb:
```diff
@@ -10,9 +10,19 @@ def self.call(*args)
+    end
 
-    MatchedContentChange.import!(columns, records)
+    MatchedContentChange.insert_all(records)
```
Sure, so this table is effective a join table between subscriber lists and content changes. So if we have two processes both trying to add to this table at around the same time they'd be adding the same joins and if they hit a duplicate then it's not likely a problem as they're just adding the same data. 

I put this in as it's an edge case that it's possible that you could have the database change after the MatchedContentChange runs and before this inserts which could then cause a crash. Thinking about it now it might be better to just crash, this would probably be more fitting with the approach we're talking about of applying an advisory lock to prevent concurrent running and is simpler than being ultra idempotent.

I previously had some tests for this on another PR today but decided they were a bit weird:

```
    it "copes in a weird concurrent situation where a different thread inserted "\
       "the matched content changes after our exist query" do
      MatchedContentChange.create!(content_change: content_change,
                                   subscriber_list: subscriber_list)
      allow(MatchedContentChange).to receive(:exists?).and_return(false)

      expect { described_class.call(content_change) }
        .to_not(change { MatchedContentChange.count })
    end
```

benthorner replied:
> if we have two processes both trying to add to this table

I'm still confused, I'm afraid. Could you help me understand the scenario when we would be trying to match the same content change at the same time? I believe this is when we'd expect a duplicate key error.

kevindew replied:
Sure, this is already replaced by the way so somewhat moot, but happy to explain.

Right now, it is possible (ie there is nothing to prevent) two process content change jobs to run at the same time.

If that does happens we probably don't want this to crash on something that isn't actually a critical problem. My expectation is that since both workers would create the same list of MatchedContentChanges then raising an error on duplicate key is an unnecessary crash because it'd just be the same data trying to be re-added.

I've come around to reason in this depth about it though is overly complex and it's probably better to just crash as the dual running isn't something the system was designed for.

kevindew commented:
Thanks Ben, changes have now been made.

> Nice one! Do we know how much faster this will be? I'm expecting it is, because we're no longer running validations. 

I've given it a few test runs locally and it came out about 25% faster. I'm expecting this effect to be a lot less in production though where the inserts are likely a lot more expensive:


```
➜  email-alert-api git:(705141e4) RAILS_ENV=test rails c
Loading test environment (Rails 6.0.3.1)
irb(main):001:0> ActiveRecord::Base.logger = nil
=> nil
irb(main):002:0> params = 10_000.times.map { |i| { subject: "subject #{i}", body: "body #{i} " * 1000, address: "email-#{i}@example.com", created_at: Time.zone.now, updated_at: Time.zone.now } }; nil
=> nil
irb(main):003:0> puts Benchmark.measure { Email.import!(params.first.keys, params.map(&:values)) }
  2.454225   0.358233   2.812458 (  4.726806)
=> nil
irb(main):004:0> puts Benchmark.measure { Email.insert_all!(params) }
  1.120689   0.316838   1.437527 (  3.301793)
=> nil
irb(main):005:0> puts Benchmark.measure { Email.import!(params.first.keys, params.map(&:values)) }
  1.895422   0.200927   2.096349 (  3.939798)
=> nil
irb(main):006:0> puts Benchmark.measure { Email.insert_all!(params) }
  0.999509   0.293736   1.293245 (  3.083494)
=> nil
irb(main):007:0> puts Benchmark.measure { Email.import!(params.first.keys, params.map(&:values)) }
  1.860550   0.261231   2.121781 (  4.276528)
=> nil
irb(main):008:0> puts Benchmark.measure { Email.insert_all!(params) }
  0.987134   0.295185   1.282319 (  3.137646)
```

> On that note, I agree that we should be pretty safe with the schema-level constraints we have; the only slight worry is if someone adds new model validations in future. I was wondering if a comment in each of the models where we rely on a lack of validations might help, with the caveat that the comment might become invalid in the future. What do you think?

Sure, I've dropped one into the affected models.

> Otherwise, my more major comments are around the repetition of the timestamp and empty array knowledge everywhere. I don't think it matters too much, and I'm just interested to explore if there's a way to DRY it up a bit before this goes in.

So I think that repetition is ok in the builders where we're very much populating on a column by column basis (though builder is probably the wrong name for them). I'm adding a method for subscription_content in https://github.com/alphagov/email-alert-api/pull/1284 and hopefully by the time you've read this I've added one to DigestRunSubscriber as I meant to do it but forgot until I started typing this reply.

kevindew commented:
> The code looks good. Thanks for looking into those comments!
> 
> This is a rather awkward point to ask this, but when did I ever let that stop me: is it easier to just use the gem? 😆 I can see advantages of using more standard APIs and avoiding a gem that may become obsolete because of them. The disadvantages appear to be the need for more boilerplate everywhere, which arguably the gem still does better than the new API.
> 
> Could you explain a bit more what motivated you to make the change?

Sure, it's a good and reasonable question. The stance I have is that by ActiveRecord introducing this feature they defined a standard way to achieve this functionality, and thus, presented with a choice it's more about does the non-standard way offer you enough advantages to justify 1) having an additional dependency and 2) using code outside the regular ActiveRecord API.

To me the differences between ActiveRecord and the gem have highlighted improvements/oddities in the app. It seems rather weird that for us to be trying to insert 1000s of records as fast as possible that we're spending time iterating them and validating every one of them (rather redundant use of validation). It also seems weird that in a number of places we're telling a record to import a collection of 0 rows into the database (which was has been a bit annoying to fix).

The biggest boilerplate seems to be the timestamps we have to add in, which is the area I've found a pain point. It's a shame the default Rails timestamps migration doesn't set current_timestamp as the default for created_at and updated_at columns as that could have all been avoided.

However despite this I think the approach ActiveRecord has taken is actually a lot closer to the functionality we want in this application. By operating at a slightly lower level we save many 1000's of object allocations, we get a modest performance increase and we remove some odd things in the code.

kevindew commented:
> > It's a shame the default Rails timestamps migration doesn't set current_timestamp as the default for created_at and updated_at columns as that could have all been avoided.
> 
> Aha, I hadn't thought of that! So could we set change the schema defaults to avoid the boilerplate?

Yup, we can do that. I'm not sure if it's quite as simple as setting it to be CURRENT_TIMESTAMP as that'll use the systems timezone, whereas Rails writes the times in UTC. It might just work though if Rails converts all times it gets out.

> I notice that the gem does make this optional. However, I agree it's a surprising default for the intended outcome of a bulk operation, so perhaps an example of where the gem was a sub-optimal abstraction. On the other hand, the default behaviour of the gem - to raise on duplicate keys - is better than the default we have now - for someone who forgets to write `!`.

I don't mind it too much myself. Seems consistent with the approach that no bang being quietly do my best (as per `create`) and with a bang being a crash the app situation. I do get your point too though.
